jax/experimental/sparse.py

```diff
@@ -0,0 +1,65 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from functools import partial
+
+import scipy.sparse.linalg
+import jax.numpy as np
+from jax.numpy.lax_numpy import _wraps
+from .. import lax
+
+_T = lambda x: np.swapaxes(np.conj(x), -1, -2)
+
+def body_fun(matvec, p, x, r, k):
+  # inspired by https://en.wikipedia.org/wiki/Conjugate_gradient_method#Example_code_in_MATLAB_/_GNU_Octave
+  r_sq = np.matmul(_T(r), r)
```

Comment:
This should be saved between iterations rather than re-computed in each step. Also, note that this quantity is exactly the same as `np.linalg.norm(r)`, so you don't need to recompute the norm in the while condition, either.